client/ionic/src/content/flow.ts

```diff
@@ -0,0 +1,114 @@
+import {
+  FLOWS_ENDPOINT,
+  FLOWS_PATH,
+  IMG_ENDPOINT,
+  IMG_PATH,
+} from './endpoints';
+import UserContext, { CONTEXT_DEFAULTS } from './userContext';
+import * as yaml from 'js-yaml';
+
+export interface BaseScreen {
+  type: string;
+  id: string;
+}
+
+// TODO: Should the texts be html instead?
+
+export interface TextImageScreen extends BaseScreen {
+  type: 'TextImage';
+  topImageUri?: string;
+  headingText: string;
+  bodyTexts?: string[];
+  bottomImageUri?: string;
+}
+
+// TODO: This is an example, and is not yet implemented.
+export interface HeroImageScreen extends BaseScreen {
+  type: 'HeroImage';
+  imageUri: string;
+}
+
+export type Screen = TextImageScreen | HeroImageScreen;
+
+export interface BaseFlow {
+  type: string;
+  screens: Screen[];
+}
+
+export interface CarouselFlow extends BaseFlow {
+  type: 'LinearCarousel';
+}
+
+export type Flow = CarouselFlow;
+
+export interface LoadedFlow {
+  readonly content: Flow;
+  readonly isLocal: boolean;
+  readonly imgPrefix: string;
+}
+
+export class FlowLoader {
+  // TODO: Disable pre-v0.1-ship if we are not confident in server support.
+  // TODO: A way to disable this easily in development.
+  static readonly ENABLE_SERVER_LOAD = true;
+
+  static readonly SUFFIX_DELIM = '.';
+
+  static contextSuffixes(ctx: UserContext): string[] {
+    // if you are in Canada using French, you would use the following content in
+    // priority order:
+    // ca.fr
+    // ca.en
+    // ww.fr
+    // ww.en
+    const suffixes = [
+      [ctx.country, ctx.language],
+      [ctx.country, CONTEXT_DEFAULTS.language],
+      [CONTEXT_DEFAULTS.country, ctx.language],
+      [CONTEXT_DEFAULTS.country, CONTEXT_DEFAULTS.language],
+    ];
+    return suffixes
+      .map(elems => elems.join(this.SUFFIX_DELIM))
+      .filter((val, idx, arr) => idx === arr.indexOf(val));
+  }
+
+  static async loadFlow(id: string, ctx: UserContext): Promise<LoadedFlow> {
+    const suffixes = this.contextSuffixes(ctx);
+    if (FlowLoader.ENABLE_SERVER_LOAD) {
+      // TODO: This is a lot of potential 404s. Consider loading a single
+      // manifest file (directory listing) that indicates which flows and
+      // suffixes exist.
+      for (const suffix of suffixes) {
+        try {
+          const url = `${FLOWS_ENDPOINT}/${id}.${suffix}.yaml`;
+          const respText = await (await fetch(url)).text();
+          return {
+            content: yaml.safeLoad(respText) as Flow,
+            isLocal: false,
+            imgPrefix: IMG_ENDPOINT,
+          };
+        } catch (e) {
+          // This is fine - we'll use the next less specific (or baked-in) flow instead.
+        }
+      }
+    }
+    for (const suffix of suffixes) {
+      try {
+        const url = `${process.env.PUBLIC_URL}/${FLOWS_PATH}/${id}.${suffix}.yaml`;
```

Comment:
I think you can probably consolidate the local/non-local into a single block of code to be a bit more DRY. e.g. defining a `FLOWS_HOST` constant instead of `FLOWS_ENDPOINT` and changing this to:

```
const host = FlowLoader.ENABLE_SERVER_LOAD ? FLOWS_HOST : process.env.PUBLIC_URL;
const url = `${host}/${FLOWS_PATH}/${id}.${suffix}.yaml`;
```

You'd make similar changes to the `isLocal` and `imgPrefix`:

```
isLocal: !FlowLoader.ENABLE_SERVER_LOAD,
imgPrefix: FlowLoader.ENABLE_SERVER_LOAD ? `${IMG_HOST}/${IMG_PATH}` : IMG_PATH,
```

Reply:
Unless the point of this is that you want to fall back to using the local flow if it doesn't exist on our server... but my guess is we'd want local to take precedence over the deployed version in our dev environment

Reply:
ah wait, important clarification... is the local fetch just for development, or is that also to load the pre-loaded flows baked into the app bundle? I'm realizing it's probably the latter, and if so, then this whole comment can be ignored

Reply:
Correct it is the latter.  Local flow is for all users when server is down or unavailable.  LMK if you want further changes.

Reply:
got it, we are good! 👍